cmd/agent/app/agent.go

```diff
@@ -148,7 +149,7 @@ func run(ctx context.Context, karmadaConfig karmadactl.KarmadaConfig, opts *opti
 		return fmt.Errorf("failed to generate execution space name for member cluster %s, err is %v", opts.ClusterName, err)
 	}
 
-	controllerManager, err := controllerruntime.NewManager(controlPlaneRestConfig, controllerruntime.Options{
+	controllerManager, err := manager.NewLabelIndexerManager(controlPlaneRestConfig, controllerruntime.Options{
```

Comment:
I don't know why we establish our own manager and cache in `util/manager`. I think we only need to add indexer to this controllerManager. And I have implement the respective indexer in controllerManager in https://github.com/karmada-io/karmada/blob/master/pkg/controllers/cluster/common.go

Reply:
And here is the list function. The field `client.Client` already use a cache here. I think there is no need to establish our own cache, right? It could make the code cleaner.

https://github.com/karmada-io/karmada/blob/616033fc0e7bb5bc68d5932ed74194582c00dad6/pkg/controllers/cluster/taint_manager.go#L68-L74

Reply:
thanks for your suggestion. 

I think there are three reasons why add custom manager and cache.
There are many callers to call list func with labelSelector and the first reason is just don't want to modify too many callers. 

The second is that when controller-runtime and client-go support use multi-indexer for labelSelector then what we 
 need to do is just use original manager and cache instead of our custom types. 

The third is that it's a non-business logic. User just do what they need to do in its controller.

Reply:
Combine your views, I think we don't need a new manager and only need to establish a new cache to meet @likakuli's three reasons.Because code about cache index only exists in controller-runtime/pkg/cache, don't need to edit manager.

  We can add new cache's constructor in here
https://github.com/karmada-io/karmada/blob/7cce95f96aeb1101b70190fd1744bbefc4dd42b0/cmd/controller-manager/app/controllermanager.go#L111-L118

Reply:
Or we can directly add indexer to controllerManager using https://github.com/karmada-io/karmada/blob/master/pkg/controllers/cluster/common.go.  It's ok for me.

Reply:
But the List method still need to rewrite to avoid changes in other spaces.

Reply:
You mean we just need a custom cache to override the List func to support list by labelSelector and don't need to call `IndexFunc` to set up indexer?

Reply:
We can call IndexFunc using cache's method.Compared to calling it on manager,it does the same thing .
https://github.com/kubernetes-sigs/controller-runtime/blob/f0351217e9e026533aece74d054b23cae5441659/pkg/cache/informer_cache.go#L172-L173

Reply:
Got it👌. We don't want to modify so many `List` callers, so we build our own cache which implement the `List` function.

Actually, I'm not sure about how much performace it will improve? I think all objects are in cache, and it seems like there have not been any perform issue so far. So is it worthy to implement our own cache and manager here? Just some questions.

Reply:
Our own cache will add index of some specific labels to optimize the List from O(n) to O(1).You can see https://github.com/kubernetes-sigs/controller-runtime/blob/f46919744bee01060c9084a285e049afffd38c9d/pkg/cache/internal/cache_reader.go#L135-L162
When List with labelselector,it use range and costs O(n).We now convert labelselector to filedselector for optimization.

Reply:
@Garrybest Do you think it is worth optimizing now? I want to finish it.

Reply:
I think it's necessary. What we need to do is thinking about when and how to do it. Also i think we should provide some best practice for users just like k8s.  How many clusters we can support? How many work we can support? e.g. Optimizing may help to increase the cap.

Reply:
Glad to see this feature. Meanwhile, I hope we could change the basic cache or manager as little as possible. 

cc @RainbowMango who may have some ideas about this patch.